### ISSUE TITLE
Auto indention does not work for parentheses

### ISSUE THREAD

**Issue by jd4no** [DonJayamanne/pythonVSCode#306]
When continuing a function call on a new line, auto indention does not recongnize parentheses as an indention point. Rather it indents back on the line's indention point.
This:
![screen shot 2016-09-12 at 5 14 37 pm](https://cloud.githubusercontent.com/assets/15043675/18454929/bfb06e5e-790c-11e6-8042-f922258813ee.png)
Should be like this:
![screen shot 2016-09-12 at 5 14 51 pm](https://cloud.githubusercontent.com/assets/15043675/18454932/c250b826-790c-11e6-8d81-e3a95475ad30.png)

**Comment by DonJayamanne** [comment]
@jd4no , that's merely a coding style, hence not necessary. 
You'd need to use a formatter to format it based on your  chosen styles (yapf or autopep8). Formatting is supported by the extension.

**Comment by gandhis1** [comment]
Just my two cents - it's a coding style, but in my experience this is how most Python is coded. Most other Python IDEs *do* indent back to the parenthesis. This is especially important in Python because of PEP-8 style guide.

if you are against having this as default behavior, I really, really think a toggle should be available.

**Comment by jd4no** [comment]
I agree with @gandhis1. I have never seen a python IDE not do this as the standard. It is definitely the PEP8 style.

**Comment by DonJayamanne** [comment]
@gandhis1 , @jd4no , a PR would be welcome

**Comment by gandhis1** [comment]
I took a look but is the VS Code API even extensible enough to implement this? I need to know the contents of the previous line to determine the location of the parenthesis, no?

![image](https://cloud.githubusercontent.com/assets/21064086/18617410/a406f92c-7d9c-11e6-96c0-baaa99a7ebc3.png)

**Comment by gandhis1** [comment]
@DonJayamanne, any update with reference to the above? 

If the code is formatted by a linter on save or on activation, this feature is not very useful. Ideally, the correct indentation needs to happen when the user presses _enter_.

**Comment by DonJayamanne** [comment]
@gandhis1 , unfortunately I haven't made any progress with this, quite simply because I don't know how to get this working at this stage.

**Comment by gandhis1** [comment]
So there isn't a way to inspect the contents of the previous line when one presses enter? Is this something that should be submitted to the vscode github as an issue?

**Comment by DonJayamanne** [comment]
@gandhis1 , you could implement the following solution:
[OnTypeFormattingEditProvider](https://code.visualstudio.com/Docs/extensionAPI/vscode-api#OnTypeFormattingEditProvider)

This gives you access to everything you need.

**Comment by DonJayamanne** [comment]
@gandhis1, I'll look into this sometime next week. Will close this idea and create a separate one's

**Comment by gandhis1** [comment]
Cool, I might take a crack at this soon, myself. Looking at that, it seems one would check for "\n", check for an open parenthesis or bracket, find the position of that character, and then add enough tabs/spaces to align it. Or is there a more robust way?

**Comment by gandhis1** [comment]
Also, to be clear, this fix would also need to indent back to the correct place after the ':' is entered, e.g. the "pass" line below:

![image](https://cloud.githubusercontent.com/assets/21064086/19837672/50275972-9e95-11e6-9499-573544dac3ad.png)

**Comment by gandhis1** [comment]
Thanks for all the excellent formatting tweaks in the latest update. 

I could probably follow exactly what you did in 393c0bb to implement this enhancement. 

Only issue is that there potentially could be an unwanted proliferation of various OnTypeFormattingProviders such that maybe it may make more sense to unite them into one? Not particularly sure whether the performance ramifications of separate providers/functions versus one would be meaningful.

**Comment by DonJayamanne** [comment]
No they have  to be separate. Besides my chance is triggered only when you type the character :

So please  proceed with your own change

**Comment by gandhis1** [comment]
Referenced the wrong commit in the previous post. Looks like the trigger is a new line?

Implementing this, I'm assuming the trigger would also be a new line.

```
// In case we have CR LF
const triggerCharacters: string[] = os.EOL.split('');
triggerCharacters.shift();
context.subscriptions.push(vscode.languages.registerOnTypeFormattingEditProvider(PYTHON, new CommentNewLineFormatProvider(), os.EOL.substring(0, 1), ...triggerCharacters));
```

**Comment by DonJayamanne** [comment]
Yes, thats right

**Comment by daicang** [comment]
I met the same problem writing Python on vscode, searched a lot, and finally found this. 

The problem using a formatter is that the formatter would correct the whole file on saving(`python.formatting.formatOnSave`), sometimes it's not your code, sometimes it is your code but you think it's okay.

So I think the best solution is providing auto indention in the plugin. @gandhis1 @DonJayamanne are you still doing this? I may help with it.

**Comment by gandhis1** [comment]
I wanted to take a crack, but i've been occupied with my day job. Feel free to go ahead. It'd be pretty much a copy of that commit with some tweaks

**Comment by pluskid** [comment]
Just want to +1 for having this feature, and for reference there is an excellent VIM plugin that does this for VIM which I have been using extensively: https://github.com/Vimjas/vim-python-pep8-indent

**Comment by adimascio** [comment]
@gandhis1 @daicang @DonJayamanne Hi,

I agree that the indentation model of python files is not quite as polished as in other editors and I'd love to improve it! I've started a branch https://github.com/adimascio/pythonVSCode/commit/a14f1f803ea8a796db0cec49027b57de47238fae as an exercise to get familiar with vscode extension API. There are still lots of things to fix, most notably

- [ ] write tests!
- [ ] handle tab-indentation mode
- [ ] fix cursor position because although the spaces are inserted, the cursor remains at the beginning of the line,

Please feel free to comment or suggest anything.

**Comment by Inch4Tk** [comment]
I fiddled around with this indentation problem a few weeks ago on my own, but never got around to posting here because I was busy. Finally had some time, so here is fairly well working version of this:
https://github.com/Inch4Tk/pythonVSCode

It can handle multiline parameter splitting. It handles ,\+-*/ at the end of line as signals to align the indent level of the next line to the indentation of the first opening (,[,{.  And separately treats hanging indents, directly after ( or [ or {.
I added an option "python.formatting.indentNewlineInParams", which is set to false by default, so it doesn't break anything without the user actually being aware what made things break ;)
Things that currently don't work:

- Tab indentation (warning inside option description)
- Cursor position if and only if the cursor would initially be placed at pos 0 after the newline. Luckily, most of the time this is not the case (e.g. inside classes or functions). This also seems very hard to fix as it looks to be an issue with vscode text insertion.

I didn't get around to writing tests, but I am not particularly motivated to do so, the feature is turned off by default anyways.
I tried to be verbose in my comments to the code, feel free to comment or suggest anything. If nothing comes up in a few days I will create a Pull request if that is okay @DonJayamanne?
